src/pudl_rmi/coordinate.py

```diff
@@ -0,0 +1,247 @@
+
+        Most of the outputs are generated via a ``execute()`` function from the
+        cooresponding module for that output. The plant-part list is generated
+        from the `pudl_out` object.
+
+        Args:
+            pudl_out (object): instance of `pudl.output.pudltabl.PudlTabl()`.
+                The frequency (`freq`) of `pudl_out` must be `AS`. For best
+                results  `fill_fuel_cost`, `roll_fuel_cost`, and `fill_net_gen`
+                should all be True.
+        """
+        self.pudl_out = pudl_out
+        if pudl_out.freq != 'AS':
+            raise AssertionError(
+                f"Frequency of `pudl_out` must be `AS` but got {pudl_out.freq}"
+            )
+
+    def get_plant_part_list(self, clobber=False):
+        """
+        Get the master unit list; generate it or get if from a file.
+
+        If you generate the MUL, it will be saved at the file path given. The
+        plant-parts list is generated via the pudl_out object.
+
+        Args:
+            clobber (boolean): True if you want to regenerate the master unit
+                list whether or not the output is already pickled. Default is
+                False.
+        """
+        if not pudl_rmi.FILE_PATH_PLANT_PARTS_EIA.is_file() or clobber:
+            logger.info(
+                "Master unit list not found "
+                f"{pudl_rmi.FILE_PATH_PLANT_PARTS_EIA} Generating a new master"
+                " unit list. This should take ~10 minutes."
+            )
+            # actually make the master plant parts list
+            plant_parts_eia = self.pudl_out.plant_parts_eia()
+            # export
+            plant_parts_eia.to_pickle(
+                pudl_rmi.FILE_PATH_PLANT_PARTS_EIA, compression='gzip'
+            )
+        else:
+            logger.info(
+                "Reading the plant part list from "
+                f"{pudl_rmi.FILE_PATH_PLANT_PARTS_EIA}"
+            )
+            plant_parts_eia = pd.read_pickle(
+                pudl_rmi.FILE_PATH_PLANT_PARTS_EIA, compression='gzip'
+            )
+        return plant_parts_eia
+
+    def get_deprish(self, clobber=False):
+        """
+        Generate of grab the cleaned deprecaition studies.
+
+        Args:
+            clobber (boolean): True if you want to regenerate the depreciation
+                data whether or not the output is already pickled. Default is
+                False.
+        """
+        if not pudl_rmi.FILE_PATH_DEPRISH.is_file() or clobber:
+            logger.info("Generating new depreciation study output.")
+            deprish_df = pudl_rmi.deprish.execute()
+            deprish_df.to_pickle(pudl_rmi.FILE_PATH_DEPRISH)
+        else:
+            logger.info(
+                "Grabbing depreciation study output from "
+                f"{pudl_rmi.FILE_PATH_DEPRISH}"
+            )
+            deprish_df = pd.read_pickle(pudl_rmi.FILE_PATH_DEPRISH)
+        return deprish_df
+
+    def get_deprish_to_eia(
```

Comment:
All of these methods have the exact same structure. I feel like there's got to be a clean way to encapsulate that instead of repeating it every time. If file exists read file and return contents. Else, generate contents, save file, and return contents. The exact instructions for "read file" and "generate contents" are different in each case, but that's all that changes, right? You could pass in two callables (a reader and a writer) that get used. Or you can infer how to read the file based on the filename (.csv, .xlsx, and .pkl.gz) and only pass in a callable that will generate the dataframe.

This pattern seems like another situation that Pydantic might be appropriate for.

Reply:
I totally agree here. I assume that we will employ a similar structure in `pudl_out` and when we do, I will use our learnings over there to clean up/standardize this object.